module/zfs/dsl_dataset.c

```diff
@@ -2425,8 +2425,9 @@ get_receive_resume_token_impl(dsl_dataset_t *ds)
 	fnvlist_free(token_nv);
 	compressed = kmem_alloc(packed_size, KM_SLEEP);
 
-	compressed_size = gzip_compress(packed, compressed,
-	    packed_size, packed_size, 6);
+	abd_t pabd = ABD_STACK_INIT(packed, packed_size);
+	compressed_size = zio_compress_data(ZIO_COMPRESS_GZIP_6,
+	    &pabd, (void **)&compressed, packed_size, 0);
```

Comment:
Mm, that's a really good point that I hadn't considered.

Above I mentioned "more refactoring" that I wanted to do later. Part of that is lifting that zero-detection out, since it's specific to the IO pipeline and not a generic compression feature.

I think I would like to keep this PR focused, so I'll change these back to direct calls (still ABD though) and reassess later.

Reply:
Reverted. I've lifted zero-detection in #16326. If/when both PRs have landed, I'll do another PR to bring this change back.